From d89c301946aa3a36428b1354ea5c79d8c0ab4a50 Mon Sep 17 00:00:00 2001
From: Ikey Doherty <ikey@solus-project.com>
Date: Thu, 8 Sep 2016 23:19:48 +0100
Subject: [PATCH] wm: Forcibly set the -rpath for the private Mutter API in
 3.21.x

This is very ugly so we'll only enable this for Mutter 3.21.0 and above,
to resolve issue #564.

"Inspiration" for this change:
https://git.gnome.org/browse/gnome-shell/commit/?id=093fd54e2b7cf2e57982f78f729394f4f1c0cf4b

Signed-off-by: Ikey Doherty <ikey@solus-project.com>
---
 configure.ac   | 15 +++++++++++++++
 wm/Makefile.am |  5 +++++
 2 files changed, 20 insertions(+)

diff --git a/configure.ac b/configure.ac
index 69b4791..9f00bae 100644
--- a/configure.ac
+++ b/configure.ac
@@ -10,6 +10,7 @@ AC_CONFIG_MACRO_DIR([m4])
 GLIB_GSETTINGS
 
 m4_define([mutter_required_version], [3.16.0])
+m4_define([mutter_goes_private_version], [3.21.0])
 m4_define([glib_required_version], [2.44.0])
 m4_define([gtk_required_version], [3.16.0])
 m4_define([peas_required_version], [1.8.0])
@@ -76,6 +77,18 @@ PKG_CHECK_MODULES([RUNDIALOG],
                 ]
 )
 
+# Check if we have the mutter private API cruft.
+PKG_CHECK_MODULES(MUTTER_PRIVATE, libmutter >= mutter_goes_private_version,
+        [have_cranky_mutter=yes], [have_cranky_mutter=no]
+)
+
+if test x$have_cranky_mutter = "xyes"; then
+        MUTTER_TYPELIB_DIR=`$PKG_CONFIG --variable=typelibdir libmutter`
+        AC_SUBST(MUTTER_TYPELIB_DIR)
+fi
+
+AM_CONDITIONAL(HAVE_CRANKY_MUTTER, test x$have_cranky_mutter = xyes)
+
 m4_include([config/configure.ac])
 
 # Got this idea from gnome-builder :)
@@ -189,6 +202,8 @@ AC_MSG_RESULT([
         stateless:              ${enable_stateless}
         xdg directory:          ${XDGDIR}
 
+        mutter private api:     ${have_cranky_mutter}
+
         Plugins:
         ========
 
diff --git a/wm/Makefile.am b/wm/Makefile.am
index 65ea954..e4123b0 100644
--- a/wm/Makefile.am
+++ b/wm/Makefile.am
@@ -27,6 +27,11 @@ budgie_wm_CFLAGS = \
 	$(DECLARATIONS) \
 	-DGNOME_DESKTOP_USE_UNSTABLE_API
 
+if HAVE_CRANKY_MUTTER
+budgie_wm_LDFLAGS = \
+	-rpath $(MUTTER_TYPELIB_DIR)
+endif
+
 budgie_wm_LDADD = \
 	${top_builddir}/config/libbudgie-config.la \
 	$(BUDGIE_BASE_LIBS) \
-- 
2.7.4

